### ISSUE TITLE
DimPlot & FindNeighbors functions produce errors

### ISSUE THREAD

**Issue by viktorfeketa** [satijalab/seurat#2841]
Hi! Please help troubleshoot some errors.
I am trying to analyze a publicly available dataset (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE113576) using the summarized output files (barcodes.tsv, genes.tsv, matrix.mtx), using the pipeline from the main PBMC tutorial, and am getting errors in DimPlot & FindNeighbors functions. All the previous functions run with no errors and give reasonable output. Using the latest versions of R and Seurat.

Errors:

```
> DimPlot(seuratObj, reduction = "pca")
Error in if (score > best$score && (!only.loose || (lmin <= dmin && lmax >=  :
  missing value where TRUE/FALSE needed
> seuratObj <- FindNeighbors(seuratObj, dims = 1:10)
Computing nearest neighbor graph
Error in validObject(r) :
  invalid class “dgTMatrix” object: all column indices (slot 'j') must be between 0 and ncol-1 in a TsparseMatrix
```

Code:
```R
library(dplyr)
library(Seurat)
library(patchwork)
data<-Read10X(".") # folder containing barcodes.tsv, genes.tsv, matrix.mtx
seuratObj <- CreateSeuratObject(counts = data, project = "moffitt2018", min.cells = 3, min.features = 200)
seuratObj <- subset(seuratObj, subset = nFeature_RNA > 200)
seuratObj <- NormalizeData(seuratObj)
seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(seuratObj)
seuratObj <- ScaleData(seuratObj, features = all.genes)
seuratObj <- RunPCA(seuratObj, features = VariableFeatures(object = seuratObj))
DimPlot(seuratObj, reduction = "pca")
seuratObj <- FindNeighbors(seuratObj, dims = 1:10)
```

**Comment by liuxl18-hku** [comment]
```r
# library(dplyr)
library(Seurat)
# library(patchwork)
data <- Read10X("download/") # folder containing barcodes.tsv.gz, features.tsv.gz, matrix.mtx.gz
> dim(data)
[1] 27998 31299
seuratObj <- CreateSeuratObject(counts = data, project = "moffitt2018", min.cells = 3, min.features = 200)
> dim(seuratObj)
[1] 20390 31299
> seuratObj[['RNA']][1:3, 1:3]
3 x 3 sparse Matrix of class "dgCMatrix"
       AAACCTGAGATGTGGC-1 AAACCTGCACACAGAG-1 AAACCTGCACTACAGT-1
Xkr4                    .                  .                  .
Gm1992                  .                  .                  .
Rp1                     .                  .                  .

seuratObj <- subset(seuratObj, subset = nFeature_RNA > 200)
> dim(seuratObj)
[1] 20390 31299

seuratObj <- NormalizeData(seuratObj)
seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(seuratObj)
seuratObj <- ScaleData(seuratObj, features = all.genes)
seuratObj <- RunPCA(seuratObj, features = VariableFeatures(object = seuratObj))
DimPlot(seuratObj, reduction = "pca")
```

![dimplot_pca](https://user-images.githubusercontent.com/42717675/78953754-7c181b00-7b0c-11ea-9aa5-fc063077b83a.png)


```r
seuratObj_nn <- FindNeighbors(seuratObj, dims = 1:10)
> names(seuratObj@graphs)
NULL
> names(seuratObj_nn@graphs)
[1] "RNA_nn"  "RNA_snn"
```

```r
> sessionInfo()
R version 3.6.3 (2020-02-29)
Platform: x86_64-conda_cos6-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /home/liuxl18/anaconda3/envs/R363/lib/libmkl_rt.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Seurat_3.1.4.9020

loaded via a namespace (and not attached):
 [1] httr_1.4.1          tidyr_1.0.2         jsonlite_1.6.1     
 [4] viridisLite_0.3.0   splines_3.6.3       lsei_1.2-0         
 [7] leiden_0.3.3        gtools_3.8.2        assertthat_0.2.1   
[10] ggrepel_0.8.2       globals_0.12.5      pillar_1.4.3       
[13] lattice_0.20-41     glue_1.4.0          reticulate_1.15    
[16] digest_0.6.25       RColorBrewer_1.1-2  colorspace_1.4-1   
[19] cowplot_1.0.0       htmltools_0.4.0     Matrix_1.2-18      
[22] plyr_1.8.6          pkgconfig_2.0.3     tsne_0.1-3         
[25] listenv_0.8.0       purrr_0.3.3         patchwork_1.0.0    
[28] scales_1.1.0        RANN_2.6.1          gdata_2.18.0       
[31] Rtsne_0.15          tibble_3.0.0        farver_2.0.3       
[34] ggplot2_3.3.0       ellipsis_0.3.0      ROCR_1.0-7         
[37] pbapply_1.4-2       lazyeval_0.2.2      cli_2.0.2          
[40] survival_3.1-11     magrittr_1.5        crayon_1.3.4       
[43] future_1.16.0       fansi_0.4.1         nlme_3.1-145       
[46] MASS_7.3-51.5       gplots_3.0.3        ica_1.0-2          
[49] tools_3.6.3         fitdistrplus_1.0-14 data.table_1.12.8  
[52] lifecycle_0.2.0     stringr_1.4.0       plotly_4.9.2.1     
[55] munsell_0.5.0       cluster_2.1.0       irlba_2.3.3        
[58] compiler_3.6.3      rsvd_1.0.3          caTools_1.18.0     
[61] rlang_0.4.5         grid_3.6.3          ggridges_0.5.2     
[64] RcppAnnoy_0.0.16    rappdirs_0.3.1      htmlwidgets_1.5.1  
[67] igraph_1.2.5        labeling_0.3        bitops_1.0-6       
[70] npsurv_0.4-0        gtable_0.3.0        codetools_0.2-16   
[73] reshape2_1.4.3      R6_2.4.1            gridExtra_2.3      
[76] zoo_1.8-7           dplyr_0.8.5         uwot_0.1.8         
[79] future.apply_1.4.0  KernSmooth_2.23-16  ape_5.3            
[82] stringi_1.4.6       parallel_3.6.3      Rcpp_1.0.4         
[85] vctrs_0.2.4         sctransform_0.2.1   png_0.1-7          
[88] tidyselect_1.0.0    lmtest_0.9-37     
```

**Comment by viktorfeketa** [comment]
Thank you for looking into this!
I see that you are not getting an error, but I still do.
Just checked sessionInfo, and our cluster is actually using R 3.6.1 - do you think this can be the issue?

```R

> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Server 7.6 (Maipo)

Matrix products: default
BLAS/LAPACK: /gpfs/ycga/apps/hpc/software/OpenBLAS/0.3.1-GCC-7.3.0-2.30/lib/libopenblas_haswellp-r0.3.1.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] Seurat_3.1.4

loaded via a namespace (and not attached):
 [1] httr_1.4.0          tidyr_0.8.3         jsonlite_1.6
 [4] viridisLite_0.3.0   splines_3.6.1       lsei_1.2-0
 [7] leiden_0.3.3        gtools_3.8.1        Rdpack_0.11-0
[10] assertthat_0.2.1    ggrepel_0.8.1       globals_0.12.5
[13] pillar_1.4.1        lattice_0.20-38     glue_1.3.1
[16] reticulate_1.12     digest_0.6.19       RColorBrewer_1.1-2
[19] colorspace_1.4-1    cowplot_0.9.4       htmltools_0.3.6
[22] Matrix_1.2-17       plyr_1.8.4          pkgconfig_2.0.2
[25] bibtex_0.4.2        tsne_0.1-3          listenv_0.8.0
[28] purrr_0.3.2         patchwork_1.0.0     scales_1.0.0
[31] RANN_2.6.1          gdata_2.18.0        Rtsne_0.15
[34] tibble_2.1.3        ggplot2_3.1.1       ROCR_1.0-7
[37] pbapply_1.4-0       lazyeval_0.2.2      survival_2.44-1.1
[40] magrittr_1.5        crayon_1.3.4        future_1.16.0
[43] nlme_3.1-140        MASS_7.3-51.4       gplots_3.0.1.1
[46] ica_1.0-2           tools_3.6.1         fitdistrplus_1.0-14
[49] data.table_1.12.2   gbRd_0.4-11         stringr_1.4.0
[52] plotly_4.9.0        munsell_0.5.0       cluster_2.0.9
[55] irlba_2.3.3         compiler_3.6.1      rsvd_1.0.3
[58] caTools_1.17.1.2    rlang_0.3.4         grid_3.6.1
[61] ggridges_0.5.1      RcppAnnoy_0.0.16    htmlwidgets_1.3
[64] igraph_1.2.4.1      bitops_1.0-6        npsurv_0.4-0
[67] gtable_0.3.0        codetools_0.2-16    reshape2_1.4.3
[70] R6_2.4.0            gridExtra_2.3       zoo_1.8-5
[73] dplyr_0.8.1         uwot_0.1.8          future.apply_1.4.0
[76] KernSmooth_2.23-15  metap_1.1           ape_5.3
[79] stringi_1.4.3       parallel_3.6.1      Rcpp_1.0.1
[82] sctransform_0.2.1   png_0.1-7           tidyselect_0.2.5
[85] lmtest_0.9-37
```

**Comment by liuxl18-hku** [comment]
Don't worry. I am not quite sure the R version, seems that you are using same version of Seurat with me (3.1.4, thought mine is develop version). 

Maybe you can do a quick check of `seuratObj[['RNA']][1:3, 1:3]` to see the data is loaded correctly or not. 

The thing is when I tried to put `genes.tsv.gz` (rename after downloaded) under download/ folder, `Seurat::Read10X()` complains, I have to change it to `features.tsv.gz`.

bless~

**Comment by satijalab** [comment]
This is almost certainly not an R version issue, but just a basic problem in how the data is being read-in. Make sure you can recapitulate the intermdiate steps that liuxl suggested. If you can confirm that the data is being read-in properly, and the issue only arises with DimPlot, please feel free to re-open this issue

**Comment by viktorfeketa** [comment]
I have tested this on my local computer with R 3.6.3 and everything works. So maybe it is in fact related to R version, or other packages associated with it (the updating from 3.6.1 to 3.6.3 on my machine was not straightforward - several packages, such as reshape2 and multtest had to be updated manually; also, there might be more differences beyond the R version - I made it work so far on Windows machine).

On R 3.6.1, I confirm that data is being read-in properly, because all of the intermediate steps produce the same output as liuxl obtained. But I am getting errors in DimPlot and FindNeighbors.

But the bottom line is that it works in R 3.6.3, so I will keep the issue closed, thanks for helping!

I noticed another minor quirk with input file names (also reported above by liuxl): if files are gunzipped, it accepts genes.tsv; however, if input files are gzipped, than genes.tsv.gz must be renamed to features.tsv.gz - maybe something you might fix to make it consistent. Thanks!
